runtime/src/current_sandbox.c

```diff
@@ -88,20 +103,43 @@ current_sandbox_start(void)
 	current_sandbox_disable_preemption(sandbox);
 	sandbox->completion_timestamp = __getcycles();
 
-	/* Retrieve the result, construct the HTTP response, and send to client */
-	if (sandbox_send_response(sandbox) < 0) {
-		error_message = "Unable to build and send client response\n";
-		goto err;
-	};
 
-	http_total_increment_2xx();
+	if (next_module != NULL) {
+		/* Generate a new request, copy the current sandbox's output to the next request's buffer, and put it to the global queue */
+		ssize_t output_length = sandbox->request_response_data_length - sandbox->request_length;
+		char * pre_func_output = (char *) malloc(output_length);
+		memcpy(pre_func_output, sandbox->request_response_data + sandbox->request_length, output_length);
+	 	uint64_t enqueue_timestamp = __getcycles();	
+		struct sandbox_request *sandbox_request =
+                                  sandbox_request_allocate(next_module, false, sandbox->request_length, 
+							   next_module->name, sandbox->client_socket_descriptor,
+                                                           (const struct sockaddr *)&sandbox->client_address,
+                                                           sandbox->request_arrival_timestamp, enqueue_timestamp, 
+							   true, pre_func_output, output_length);
+		/* TODO: all sandboxs in the chain share the same request id, but sandbox_request_allocate() will busy-wait to generate an unique
+		   id, should we optimize it here?*/
+		sandbox_request->id = sandbox->id;  
+		/* Add to the Global Sandbox Request Scheduler */
+		global_request_scheduler_add(sandbox_request);
```

Comment:
I'm guessing that this code does *not* include your optimization for avoiding the global queue in the case that we can run the next sandbox on our local core.

@bushidocodes @emil916 Should we expect any problems with workers adding onto the global queue in *addition* to the listener core?

Reply:
> I'm guessing that this code does _not_ include your optimization for avoiding the global queue in the case that we can run the next sandbox on our local core.
> 
> @bushidocodes @emil916 Should we expect any problems with workers adding onto the global queue in _addition_ to the listener core?

Yes, this code doesn't include the optimization part for avoiding the global queue, I still have some concerns for this but I will define a macro to control this optimization.

Reply:
I believe we should not expect any problems, as long as we have LOCKs enabled to access the global queue, then workers technically can add to it with no problem.
I don't get line 136 much, is this for making sure the 2nd and the 3rd inner functions of the chain get handled by the same worker? But even if so, how will the different sandboxes have the same exact absolute deadlines? I must be missing something here.

Reply:
I wrote the global request queue minheap variant specifically assuming that it's only called by the listener and panics out the program if it's called by a worker. I assume that was changed in this PR for this to function.

Reply:
> I believe we should not expect any problems, as long as we have LOCKs enabled to access the global queue, then workers technically can add to it with no problem.
> I don't get line 136 much, is this for making sure the 2nd and the 3rd inner functions of the chain get handled by the same worker? But even if so, how will the different sandboxes have the same exact absolute deadlines? I must be missing something here.

You are right, line 136 tries to check if the current sandbox has the same deadline with the next sandbox request or not, if they are the same, the worker thread will put the next sandbox(after initializing it) into its local runqueue not the global queue, this is to avoid the global queue overhead. But based on my performance tests, this optimization seems not to improve the performance, so I add a macro to control this feature, the default setting for this feature is off.